### PR TITLE
feat: send emails via http api endpoint instead of smtp

### PR DESCRIPTION
**Pull request by barnarddt**
This change adds a new delivery method to the courier called `mailer`. Similar to SMS functionality it posts a templated Data model to a API endpoint.  This API can then send emails via a CRM or any other mechanism that it wants.

`Mailer` still uses the existing email data models so any new email added will automatically be sent to the API/CRM as well.

## Related issue(s)
Resolves https://github.com/ory/kratos/issues/2825
Also see #1030 and #3008 
Documentation PR https://github.com/ory/docs/pull/1298

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

**Comment by karaimin** [comment]
Nice work @barnarddt. I was looking for this feature for a long time. Hope it will be reviewed and released soon.

File: courier/stub/request.config.mailer.jsonnet [hunk 1]
```diff
@@ -0,0 +1,11 @@
+function(ctx) {
```
Comment (aeneasr) [inline courier/stub/request.config.mailer.jsonnet:1]: This looks like a good default value for the body payload. Maybe it makes sense to use it as the default?

File: driver/config/config.go [hunk 2]
```diff
@@ -78,6 +78,8 @@ const (
 	ViperKeyCourierTemplatesVerificationValidEmail           = "courier.templates.verification.valid.email"
 	ViperKeyCourierTemplatesVerificationCodeInvalidEmail     = "courier.templates.verification_code.invalid.email"
 	ViperKeyCourierTemplatesVerificationCodeValidEmail       = "courier.templates.verification_code.valid.email"
+	ViperKeyCourierMailerEnabled                             = "courier.mailer.enabled"
```
Comment (aeneasr) [inline driver/config/config.go:81]: I would prefer this layout:

```
courier:
  email:
    strategy: smtp|http
  http:
    url: ...
  smtp:
    connection_uri: ...
```

**Review by aeneasr — COMMENTED** [review]
Thank you for the PR! This is already going in a good direction. Here are some improvement ideas:

1. Add tests for the new config options (right now the config won't work because it's not in the config JSON Schema)
2. Add an e2e tests to test/e2e which ensures that this works end-to-end
3. Make a few changes to the configuration layout
4. Add documentation :)

File: courier/stub/request.config.mailer.jsonnet (hunk 1 above)
Comment (barnarddt) [reply courier/stub/request.config.mailer.jsonnet:1]: Not sure how we'd make it the default as the body of the template engine is a file URI not a absolute value?

**Comment by barnarddt** [comment]
> 2\. Add an e2e tests to test/e2e which ensures that this works end-to-end

@aeneasr 

For the end2end test what would be the preferred setup? Configure kratos to use the http email functionality with the url being a localhost port that the end2end test will listen on and ensure that that gets called when a admin call gets issued to do something that will trigger an email to be sent?

**Comment by SoGoDev** [comment]
Guys, any updates on this PR?

File: courier/courier.go [hunk 3]
```diff
@@ -49,11 +49,12 @@ type (
...
 
 	courier struct {
-		smsClient   *smsClient
-		smtpClient  *smtpClient
-		deps        Dependencies
-		failOnError bool
-		backoff     backoff.BackOff
+		smsClient    *smsClient
+		smtpClient   *smtpClient
+		mailerClient *mailerClient
```
Comment (jonas-jonas) [inline courier/courier.go:54]: Could we name this `httpClient` instead? The `smtpClient` also refers to the protocol it's using, and `mailer` is arguably a bit ambiguous.

File: courier/mailer.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/ory/kratos/request"
+)
+
+type mailerDataModel struct {
+	Recipient    string
+	TemplateType TemplateType
+	TemplateData EmailTemplate
+}
+
+type mailerClient struct {
```
Comment (jonas-jonas) [inline courier/mailer.go:22]: And similarly here.

**Review by jonas-jonas — CHANGES_REQUESTED** [review]
This already looks fantastic. And sorry for the delays on our side. 

As for the E2E tests, we already have some "mock" services that just provide a target for the webhooks to be tested with, here: https://github.com/ory/kratos/blob/master/test/e2e/mock. That service is started in [test/e2e/modd.conf](https://github.com/ory/kratos/blob/master/test/e2e/modd.conf)

Maybe you could set up some kind of echo server here, that makes all incoming messages available on an endpoint and assert that the expected messages end up in there?

File: courier/courier.go (hunk 3 above)
Comment (kmherrmann) [reply courier/courier.go:54]: +1 ... I thought the same

**Comment by kmherrmann** [comment]
Thanks @barnarddt - this is a great addition. Would you be up to address the requested changes? It'd be fantastic to get http email delivery support merged!

Resolves #2825

**Review by aeneasr — COMMENTED** [review]
Great job! :)

Could you please add an end-to-end Testcafe / Cypress test (or some other type of integration test in Go) to prove that this does indeed work as intended? Thank you!

**Comment by aeneasr** [comment]
> > 2. Add an e2e tests to test/e2e which ensures that this works end-to-end
> 
> @aeneasr
> 
> For the end2end test what would be the preferred setup? Configure kratos to use the http email functionality with the url being a localhost port that the end2end test will listen on and ensure that that gets called when a admin call gets issued to do something that will trigger an email to be sent?

Sorry - I missed this comment :(

Yes that could work! We now support PlayWright and there you can set up a mock server easily: https://playwright.dev/docs/test-webserver

So probably the test would be this:

1. set up web server
2. trigger registration flow
3. ensure web server got api request with correct payloads (potentially using a snapshot)

That should be it!

**Comment by hperl** [comment]
@barnarddt thank you so much for your contribution! I did some clean-ups in https://github.com/ory/kratos/pull/3341 to get this across the finish line:

* added the config values to the schema, including a default configuration for the `request_config`
* added an E2E test
* added subject and body to the template directly

**Comment by hperl** [comment]
Closing in favor of https://github.com/ory/kratos/pull/3341. Again, big thanks to @barnarddt for the excellent work!

**Pull request closed without merging**